### PR TITLE
[Feature] Introduce Flink CDC service

### PR DESCRIPTION
**Pull request by hdygxsj**
### Purpose

Defined the ActionService api and added an implementation for the mysql_sync_table CDC job.

Next, all that's needed is to parse the graph into an ActionContext object, and then call the ActionService with it.

For example:
```java
        ActionService actionService = new FlinkCdcActionService();
        ActionContext actionContext =
                MysqlSyncTableActionContext.builder()
                        .sessionUrl("xxxx")
                        .flinkJobType(FlinkJobType.SESSION)
                        .actionPath(
                                "/xxxx")
                        .warehouse("s3://paimon/cdc")
                        .database("paimonweb")
                        .table("catalog")
                        .primaryKeys("id")
                        .mysqlConfList(
                                Arrays.asList(
                                        "hostname=xxxx",
                                        "port=30001",
                                        "username=xxx",
                                        "password=xxxx",
                                        "database-name=paimon",
                                        "table-name=catalog",
                                        "server-time-zone=Asia/Shanghai"))
                        .catalogConfList(
                                Arrays.asList(
                                        "s3.endpoint=http://xxxx",
                                        "s3.access-key=xxxx",
                                        "s3.secret-key=xxxx"))
                        .tableConfList(
                                Arrays.asList(
                                        "bucket=4",
                                        "changelog-producer=input",
                                        "sink.parallelism=4"))
                        .build();
        actionService.execute(actionContext);
```

### Tests

- `FlinkCdcActionContextTestBase`
- `MysqlSyncDatabaseActionContextTest`
- `MysqlSyncTableActionContextTest`
- `PostgresSyncTableActionContextTest`

### API and Format

- `ActionContext`
```java
/** The context of action which converts the user-defined action to command line. */
public interface ActionContext {
    String name();

    /** Converts the user-defined action to command line. */
    List<String> getArguments();

    String getJarPath();
}

```

- `ActionService`
```java
/**
 * Action service definition. Convert the user-configured CDC job into an ActionContext {@link
 * ActionContext}, and then invoke the execute method to execute the action job.
 */
public interface ActionService {

    ActionExecutionResult execute(ActionContext actionContext) throws Exception;
}

```
### Documentation

No.

### CODE REVIEW

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.paimon.web.api.action.context.ActionContext;
+import org.apache.paimon.web.api.exception.ActionException;
+import org.apache.paimon.web.api.shell.ShellService;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+
+/** Action service. */
+public interface ActionService {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java:30]: Add the defailed description of `ActionService` interface.

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    default String getFlinkHome() {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java:32]: Is the `getFlinkHome` need to define as the interface of `ActionService`?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        if (StringUtils.isBlank(flinkHome)) {
+            flinkHome = System.getProperty("FLINK_HOME");
+        }
+        if (StringUtils.isBlank(flinkHome)) {
+            throw new ActionException("FLINK_HOME is null");
+        }
+        return flinkHome;
+    }
+
+    default String getActionPath() {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java:43]: Ditto.

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        if (StringUtils.isBlank(actionPath)) {
+            throw new ActionException("ACTION_PATH is null");
+        }
+        return actionPath;
+    }
+
+    String name();
+
+    default void execute(ActionContext actionContext) throws Exception {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java:56]: Why does the default implementation of `execute` interface use Flink?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            Process process = new ShellService(flinkHome, getCommand(actionContext)).execute();
+            afterExecute(process, actionContext);
+        } catch (Exception exception) {
+            catchException(exception);
+        }
+    }
+
+    List<String> getCommand(ActionContext actionContext);
+
+    default void beforeExecute(ActionContext actionContext) {}
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java:69]: Why does this need to define the execute series interfaces?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        beforeExecute(actionContext);
+        try {
+            Process process = new ShellService(flinkHome, getCommand(actionContext)).execute();
+            afterExecute(process, actionContext);
+        } catch (Exception exception) {
+            catchException(exception);
+        }
+    }
+
+    List<String> getCommand(ActionContext actionContext);
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/ActionService.java:67]: Could you add some comment for `getCommand`?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/MysqlSyncTableActionService.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.apache.paimon.web.api.action.context.ActionContext;
+
+import com.google.auto.service.AutoService;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Mysql sync table action service . */
+@AutoService(ActionService.class)
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/MysqlSyncTableActionService.java:29]: Why does this use auto service of google?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/MysqlSyncTableActionService.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+public class MysqlSyncTableActionService implements ActionService {
+
+    @Override
+    public String name() {
+        return "mysql_sync_table";
+    }
+
+    @Override
+    public List<String> getCommand(ActionContext actionContext) {
+        List<String> commandList = new ArrayList<>();
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/MysqlSyncTableActionService.java:39]: Why does `flink run` return multiple command? It's so confusing for developers.

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.paimon.web.api.action.context;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+
+/** Action Context. */
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java:25]: Add the detailed comment for the `ActionContext` interface.

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+public interface ActionContext {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java:26]: Please redesign the public interfaces of `ActionContext`. The definition of these interfaces do not look good to me.

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/MysqlSyncTableActionContext.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+    private List<String> mysqlConfList;
+
+    private List<String> catalogConfList;
+
+    private List<String> tableConfList;
+
+    private MysqlSyncTableActionContext() {}
+
+    public static MysqlSyncTableActionContext builder() {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/MysqlSyncTableActionContext.java:49]: Is there any definition of builder for `MysqlSyncTableActionContext`?

**Review by SteNicholas — CHANGES_REQUESTED** [review]
@hdygxsj, thanks for the contribution. I left some comments for this pull request. BTW, could you firstly give a clear definition of the introduced public interface?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/MysqlSyncTableActionService.java (hunk 8 above)
Comment (hdygxsj) [reply paimon-web-api/src/main/java/org/apache/paimon/web/api/action/MysqlSyncTableActionService.java:39]: Because calling a shell requires converting the command into a List.
image

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/MysqlSyncTableActionService.java (hunk 7 above)
Comment (hdygxsj) [reply paimon-web-api/src/main/java/org/apache/paimon/web/api/action/MysqlSyncTableActionService.java:29]: The intention is to make this feature pluggable, enabling users to upload their own JAR files in the future and thereby implement custom Action tasks according to their specific needs.

File: paimon-web-api/pom.xml [hunk 12]
```diff
@@ -120,5 +120,9 @@ under the License.
...
             <version>${paimon.version}</version>
         </dependency>
+        <dependency>
+            <groupId>com.google.auto.service</groupId>
+            <artifactId>auto-service</artifactId>
+        </dependency>
     </dependencies>
-</project>
\ No newline at end of file
+</project>
```
Comment (SteNicholas) [inline paimon-web-api/pom.xml:128]: Revert this unnecessary change.

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.web.api.action.context;
+
+import java.util.List;
+
+/**
+ * Action Context. Convert the user-defined CDC job to an ActionContext and then to a command line
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java:24]:
```suggestion
 * The context of action which converts the user-defined job to a command line.
```

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+ * argument
+ */
+public interface ActionContext {
+
+    List<String> getCommand();
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java:29]: Add comment for this method.

**Review by SteNicholas — CHANGES_REQUESTED** [review]

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/MysqlSyncTableActionContext.java [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.paimon.web.api.action.context.ActionContextUtil.addConf;
+import static org.apache.paimon.web.api.action.context.ActionContextUtil.addConfList;
+
+/** Mysql sync table action context. */
+@AllArgsConstructor
+public class MysqlSyncTableActionContext implements ActionContext {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/MysqlSyncTableActionContext.java:34]: Use `lombok.Builder` for builder.

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/MysqlSyncTableActionContext.java [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+    private MysqlSyncTableActionContext() {}
+
+    public static MysqlSyncTableActionContext.Builder builder() {
+        return new MysqlSyncTableActionContext.Builder();
+    }
+
+    @Override
+    public List<String> getCommand() {
+        List<String> command = new ArrayList<>();
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/MysqlSyncTableActionContext.java:60]: Move this line after lne 63.

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/MysqlSyncTableActionContext.java [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+        if (StringUtils.isAnyBlank(warehouse, database, table)) {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/MysqlSyncTableActionContext.java:61]: Why is the check whether is blank of `warehouse`, `database` and `table` in `getCommand` not contructor?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/AbstractActionService.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/** An abstract Action service that executes actions through the shell. */
+@Slf4j
+public abstract class AbstractActionService implements ActionService {
+
+    public List<String> getCommand(ActionContext actionContext) {
+        List<String> commandList = new ArrayList<>();
+        commandList.add("${FLINK_HOME}/bin/flink");
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/AbstractActionService.java:37]: Why does the base implementation of `getCommand` is based on Flink command?

File: paimon-web-api/src/test/java/org/apache/paimon/web/api/action/MysqlSyncTableActionServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    private String getActionPath() {
+        return actionPath.toUri().toString();
+    }
+
+    private String getFlinkHome() {
+        return flinkHomePath.toUri().toString();
+    }
+
+    @Test
+    public void testBuildCommand() {
```
Comment (SteNicholas) [inline paimon-web-api/src/test/java/org/apache/paimon/web/api/action/MysqlSyncTableActionServiceTest.java:69]:
```suggestion
    public void testGetCommand() {
```

**Review by SteNicholas — CHANGES_REQUESTED** [review]

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/MysqlSyncTableCdcActionService.java [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.web.api.action.service;
+
+import com.google.auto.service.AutoService;
+
+/** Mysql sync table action service . */
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/MysqlSyncTableCdcActionService.java:23]:
```suggestion
/** Mysql sync table action service. */
```

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/AbstractCdcActionService.java [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+        if (StringUtils.isBlank(actionPath)) {
+            actionPath = System.getProperty("ACTION_PATH");
+        }
+        if (StringUtils.isBlank(actionPath)) {
+            throw new ActionException("ACTION_PATH is null");
+        }
+        return actionPath;
+    }
+
+    public void catchException(Exception exception) throws Exception {}
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/AbstractCdcActionService.java:77]: What's the purpose of `catchException`?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/AbstractCdcActionService.java [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** An abstract Action service that executes actions through the shell. */
+@Slf4j
+public abstract class AbstractCdcActionService implements ActionService {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/AbstractCdcActionService.java:33]:
```suggestion
public abstract class FlinkCdcActionService implements ActionService {
```

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/AbstractCdcActionService.java [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.paimon.web.api.exception.ActionException;
+import org.apache.paimon.web.api.shell.ShellService;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** An abstract Action service that executes actions through the shell. */
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/service/AbstractCdcActionService.java:31]:
```suggestion
/** An Flink CDC Action service that executes actions through the shell. */
```

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.apache.paimon.web.api.action.context;
+
+import java.util.List;
+
+/** The context of action which converts the user-defined action to command line. */
+public interface ActionContext {
+
+    /** Converts the user-defined action to command line. */
+    List<String> getCommand();
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java:27]:
```suggestion
    String getCommand();
```

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionConf.java [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+/** ActionConf. */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.FIELD)
+public @interface ActionConf {
+
+    String value();
+
+    boolean nullable() default true;
+
+    boolean confList() default false;
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionConf.java:35]: Remove this unnecessary interface.

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/enums/ActionExecuteResult.java [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.web.api.enums;
+
+/** ActionExecuteResult. */
+public enum ActionExecuteResult {
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/enums/ActionExecuteResult.java:22]:
```suggestion
public enum ActionExecutionResult {
```

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+/** The context of action which converts the user-defined action to command line. */
+public interface ActionContext {
+    String name();
+
+    /** Converts the user-defined action to command line. */
+    List<String> getActionArgs();
+
+    String getActionJarPath();
+
+    String getErrorMessage();
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java:34]: Could this interface merge with `getExecutionResult`?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+
+    ActionExecuteResult getExecuteResult();
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java:36]:
```suggestion
    ActionExecuteResult getExecutionResult();
```

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.List;
+
+/** The context of action which converts the user-defined action to command line. */
+public interface ActionContext {
+    String name();
+
+    /** Converts the user-defined action to command line. */
+    List<String> getActionArgs();
+
+    String getActionJarPath();
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java:32]:
```suggestion
    String getJarPath();
```

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.paimon.web.api.enums.ActionExecuteResult;
+
+import java.util.List;
+
+/** The context of action which converts the user-defined action to command line. */
+public interface ActionContext {
+    String name();
+
+    /** Converts the user-defined action to command line. */
+    List<String> getActionArgs();
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionContext.java:30]:
```suggestion
    List<String> getArguments();
```

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/enums/FlinkJobType.java [hunk 31]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.web.api.enums;
+
+/** FlinkJobType. */
+public enum FlinkJobType {
+    SESSION,
+    JOB,
```
Comment (SteNicholas) [inline paimon-web-api/src/main/java/org/apache/paimon/web/api/enums/FlinkJobType.java:24]:
```suggestion
    PER_JOB,
```

**Review by SteNicholas — CHANGES_REQUESTED** [review]

**Comment by SteNicholas** [comment]
@hdygxsj, thanks for updates. Could you also update the description of this pull request according to new introduced interfaces?

File: paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionConf.java (hunk 25 above)
Comment (hdygxsj) [reply paimon-web-api/src/main/java/org/apache/paimon/web/api/action/context/ActionConf.java]: They are utilized for null value evaluations and the construction of arguments.
image
image

**Review by SteNicholas — APPROVED** [review]
LGTM.
